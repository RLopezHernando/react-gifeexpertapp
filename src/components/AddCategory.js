import React from 'react'
import PropTypes from 'prop-types';
import { useState } from 'react';

//En () nos vienen las props enviadas al hijo desde el padre
//Normalmente estas props se desestructuran y se pone ({setCategories})
// Si no desestructuramos serian props en generico y luego props.setCategories

export const AddCategory = ({setCategories}) => {
    const [inputValue, setInputValue] = useState('');
    const handleInputChange = (e) => {
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault();

        if(inputValue.trim().length >2){
            setCategories(cats => [inputValue,...cats]);
            setInputValue('');
        }
    }

    return (

        <form onSubmit={handleSubmit}>
            <p>{inputValue}</p>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>


    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
