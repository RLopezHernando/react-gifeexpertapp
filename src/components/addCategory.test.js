import '@testing-library/jest-dom'
import { AddCategory } from './AddCategory';
import {shallow,wrapper} from 'enzyme';
describe('Pruebas eb <AddCategory />', () =>{

    // const setCategories = () =>{};
    //Para llamar a una funcion con jest
    const setCategories = jest.fn();    
    let wrapper  = shallow(<AddCategory setCategories = {setCategories}/>);

    // Antes de la ejecuccion de las pruebas se da esto. Es la precarga antes de las pruebas
    beforeEach( () =>{  
        jest.clearAllMocks();
        //Reseteamos el componente. Se ejecuta esto antes que las pruebas al estar en el beforeEach();
        wrapper = shallow(<AddCategory setCategories = {setCategories}/>);
    })


    test('Debe de mostrarse correctmente',() =>{
        
        expect(wrapper).toMatchSnapshot();
    })


    test('Debe de cambiar la caja de texto',() =>{
        const input = wrapper.find('input');
        const value = 'Hola Mundo';
        input.simulate('change',{
            target:{value}
        });
        expect(wrapper.find('p').text().trim()).toBe(value);
        
    })

    test('No debe de postear la informacion on submit',() =>{
        //Simulamos el submit
        wrapper.find('form').simulate('submit',{
            preventDefault:() =>{}
        })
        //Si no hay texto en el imput no se llama la funcion
        expect(setCategories).not.toHaveBeenCalled();
    })

    test('Debe de llamar el setCategories y limpiar la caja de texto',() => {
        const value = "Hola Mundo"
        // 1. Simular el inputChange
        wrapper.find('input').simulate('change',{target:{value}});
        // 2. Simular el submit
        wrapper.find('form').simulate('submit',{
            preventDefault:() =>{}
        })
        //3. setCategories se debe haber llamado
        expect(setCategories).toHaveBeenCalled();
        expect(setCategories).toHaveBeenCalledTimes(1);
        expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

        //4. el valor del input debe de estar ''
        expect(wrapper.find('input').prop('value')).toBe('')
    })
})

