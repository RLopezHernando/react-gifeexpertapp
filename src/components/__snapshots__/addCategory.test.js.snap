// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas eb <AddCategory /> Debe de mostrarse correctmente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
